content/blog/zephyr-tracing-and-profiling/index.md

```diff
@@ -0,0 +1,279 @@
+
+The tracing format CTF requires data typing to be decoded, so we’ll copy the metadata (tracing format description) file to the same folder:
+
+```shell
+cp subsys/tracing/ctf/tsdl/metadata ../tracing/
+```
+
+Finally, to parse the CTF data and show the actual trace:
+
+```shell
+babeltrace2 ../tracing/
+```
+
+The output is something like:
+
+<a href="img/img001.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img001.webp"
+        alt=""
+        caption="Sample trace output"
+    >}}
+</a>
+
+This log shows the execution order and timing of the traced objects, allowing us to visualize how the software is executing in the acquired interval. This gives the possibility of running statistics on how the CPU resources are used and easily spot functions that are taking too long to execute, are being executed at wrong moments or more times than expected.
+
+### Objects monitored
+
+It is possible to control which kinds of objects are monitored. The following are currently supported [[1]]:
+
+```
+struct k_timer *_track_list_k_timer;
+struct k_mem_slab *_track_list_k_mem_slab;
+struct k_sem *_track_list_k_sem;
+struct k_mutex *_track_list_k_mutex;
+struct k_stack *_track_list_k_stack;
+struct k_msgq *_track_list_k_msgq;
+struct k_mbox *_track_list_k_mbox;
+struct k_pipe *_track_list_k_pipe;
+struct k_queue *_track_list_k_queue;
+struct k_event *_track_list_k_event;
+```
+
+Kconfig’s menu *Tracing Configuration* will allow to select which objects will be traced, with options like shown below:
+
+```
+TRACING_SYSCALL=y/n
+TRACING_ISR=y/n
+TRACING_SEMAPHORE=y/n
+TRACING_HEAP=y/n
+TRACING_STACK=y/n
+TRACING_PM=y/n
+(...)
+```
+
+If a large amount of data is to be generated, this might be useful to limit tracing only to the objects of interest, both to save buffer and provide a cleaner data acquisition.
+
+Another interesting feature is the possibility of defining user functions to be called at tracing events, as *headers* shown below. In a debugging situation or when external equipment (e.g oscilloscope or data logger) is used, it can be useful to add custom code to the callbacks to drive a GPIO, for instance, in order to sync external acquisition data with the log provided by the *tracing* subsystem.
+
+<a href="img/img002.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img002.webp"
+        alt=""
+        caption="Trace user functions"
+    >}}
+</a>
+
+The following macros [[2]] are placed on Zephyr’s subsystems code to track important functions of the operating system:
+
+```
+SYS_PORT_TRACING_FUNC(type, func, …)
+SYS_PORT_TRACING_FUNC_ENTER(type, func, …)
+SYS_PORT_TRACING_FUNC_EXIT(type, func, ...)
+```
+
+This means that depending on the functionality being probed, the *tracing* subsystem might already have hooks that will acquire data by default, as can be seen in the example below:
+
+```c

```

Comment:
Sounds there is issues with the formatting here. Try removing the `c`.

Reply:
I see no issues.
<img width="685" alt="Screenshot 2024-11-04 at 09 13 42" src="https://github.com/user-attachments/assets/b7a2ddc4-5cdf-4e84-977d-558b83d10168">

Reply:
I couldn't notice problems as well